src/Illuminate/Redis/Lua/LuaScript.php

```diff
@@ -0,0 +1,237 @@
+     * @return self
+     */
+    public static function fromSha1($connection, $sha1)
+    {
+        return new static($connection, $sha1, null, false);
+    }
+
+    /**
+     * The boolean flag that indicates whether the script should be cached or not.
+     *
+     * @return bool
+     */
+    public function isCachingEnable()
+    {
+        return $this->isCachingEnable;
+    }
+
+    /**
+     * Get the list of the keys to be passed to the script.
+     *
+     * @return array
+     */
+    public function getKeys()
+    {
+        return $this->keys;
+    }
+
+    /**
+     * Get the list of the arguments to be passed to the script.
+     *
+     * @return array|null
+     */
+    public function getArguments()
+    {
+        return $this->arguments;
+    }
+
+    /**
+     * Get the lua script code.
+     *
+     * @return string|null
+     */
+    public function getScript()
+    {
+        return $this->script;
+    }
+
+    /**
+     * Get the SHA1 hash of the script that's already loaded in Redis.
+     *
+     * @return string|null
+     */
+    public function getSha1()
+    {
+        return $this->sha1;
+    }
+
+    /**
+     * Set the list of keys and arguments to be passed to the script.
+     *
+     * @param  array  $keys  The list of keys to be passed to the script.
+     * @param  array|null  $arguments  The list of arguments to be passed to the script.
+     * @return $this
+     */
+    public function with($keys, $arguments = null)
+    {
+        $this->keys = $keys;
+        $this->arguments = $arguments;
+
+        return $this;
+    }
+
+    /**
+     * Execute the Lua script with the given SHA1 hash using the given Redis connection.
+     *
+     * @param  string  $sha  The SHA1 hash of the Lua script.
+     * @return mixed The result of the script execution.
+     */
+    protected function executeWithHash($sha)
+    {
+        return $this->connection->command('evalsh', [$sha, $this->keys, $this->arguments, count($this->keys)]);
+    }
+
+    /**
+     * Execute the Lua script using EVAL command with the given Redis connection.
+     *
+     * @return mixed The result of the script execution.
+     */
+    protected function executeWithScript()
+    {
+        return $this->connection->eval($this->script, count($this->keys), $this->keys, $this->arguments);
+    }
+
+    /**
+     * Execute the Lua script using the given Redis connection.
+     *
+     * @return mixed The result of the script execution.
+     *
+     * @throws \Illuminate\Redis\Lua\Exception\LuaScriptLoadException If the lua script could not be loaded.
+     */
+    public function execute()
+    {
+        if ($this->sha1 !== null) {
+            return $this->executeWithHash($this->sha1);
+        } else {
+            if ($this->isCachingEnable) {
+                // Try to execute the script.
+                $hash = sha1($this->script);
+                $result = $this->executeWithHash($hash);
+
+                if ($result !== false) {
```

Comment:
Lua has a boolean type. Won't returned false from the Lua script (like "return false") get here as PHP false? Native radis command used to return integers, but I don't knows.

Reply:
The return value of the Lua script could be a scalar value `int`, `string`, or an `array`. Arrays that are returned can also contain other arrays. if there is an error in executing the Lua script it's it should return `false`.

Reply:
```
sail artisan tinker
Psy Shell v0.11.14 (PHP 8.2.4 — cli) by Justin Hileman
> $redis = \Illuminate\Support\Facades\Redis::connection();
= Illuminate\Redis\Connections\PhpRedisConnection {#3737}

> $redis->eval('return 2 == 2', 0);
= 1

> $redis->eval('return 2 == 3', 0);
= false

> $redis->eval('return not false', 0);
= 1

> $redis->eval('return not true', 0);
= false

> $redis->eval('return not nil', 0);
= 1

> $redis->eval('return nil', 0);
= false

> $redis->eval('redis.call("SET", "key2", "value"); redis.call("SET", "key3", "value"); return redis.call("GET", "key2") == redis.call("GET", "key3")', 0);
= 1

> $redis->eval('redis.call("SET", "key2", "value2"); redis.call("SET", "key3", "value3"); return redis.call("GET", "key2") == redis.call("GET", "key3")', 0);
= false
```
It turns out that Redis can return `false` when the script have been executed successfully. We can move out equality check on redis side (i.e. take a value fron list1, and compare it to value from list 2), we don't want to pull values to our side, we only want the result of the comparison.

Reply:
Side note. In lua `0`, as well as empty strings, means `true`. Only `false` and `nil` test false. May be, because of it redis doesn't convert false to 0, but do convert true to 1. But idk for sure.

Reply:
No prob. You are welcome.

Yes, `getLastError()`can help with it. But also `clearLastError()` required before eval call, because there is no any auto-clear with phpredis:
```
> $redis->eval('bad code', 0);
= false

> $redis->client()->getLastError();
= "ERR Error compiling script (new function): user_script:1: '=' expected near 'code'"

> $redis->eval('return 1 == 2', 0);
= false

> $redis->client()->getLastError();
= "ERR Error compiling script (new function): user_script:1: '=' expected near 'code'"

> $redis->eval('return "Hello"', 0);
= "Hello"

> $redis->client()->getLastError();
= "ERR Error compiling script (new function): user_script:1: '=' expected near 'code'"
```
It should be enough for phpredis.

But `predis` doesn't have anything like it. It throw exceptions:
```
> $redis->eval('bad code', 0);

   Predis\Response\ServerException  ERR Error compiling script (new function): user_script:1: '=' expected near 'code'.

> $e = $redis->evalsha(null, 0);

   Predis\Response\ServerException  NOSCRIPT No matching script. Please use EVAL.
```
Or return error object (if exceptions are switched of by config options):
```
> $redis->eval('bad code', 0);
= Predis\Response\Error {#3753}

> $e = $redis->evalsha(null, 0);
= Predis\Response\Error {#3759}

> $e->getMessage()
= "NOSCRIPT No matching script. Please use EVAL."
```
So, catch, check, proceed or throw further (don't know for sure, if it possible to write script that can be successfully cached, but throw an error been executed).

As far as Laravel support both drivers, this LuaScript implementation either should support both of them. We have no way to extend abstract Connection class with any lastError() method - predis doesn't have any, and catching all the errors to implement it manually will introduce a bad BC (user code can relay on exceptions).

It's a bunch of logic here to made this "fast call in hope of success". Strategy template is required.